File: mq/rabbitmq/php/src/broadcast/subscriber_printer.php
<?php
/**
 * rabbitmq hello-world
 */
require_once __DIR__ . '/../bootstrap.php';


$connection = new \PhpAmqpLib\Connection\AMQPStreamConnection(RABBITMQ_MASTER, 5672, 'guest', 'guest');
$channel = $connection->channel();

$channel->queue_declare('logs_broadcast_printer', false, true, false, true);
$channel->queue_bind('logs_broadcast_printer', 'logs_broadcast', 'logs_broadcast_printer');

$channel->basic_consume('logs_broadcast_printer', null, false, false, false, false,
    function(\PhpAmqpLib\Message\AMQPMessage $message) {
        $data = json_decode($message->body);
        echo sprintf("retrieve message <%s:%s@%s> %s\n", $data->from, $data->producer, $data->index, $data->body);

        /* @var \PhpAmqpLib\Channel\AMQPChannel $channel */
        $channel = $message->delivery_info['channel'];
        for ($index = 0; $index < random_int(1, 3); ++$index) {
            echo '.';
            sleep(1);
            if (random_int(1, 25) === 1) {
                echo "reject\n";
                return $channel->basic_reject($message->getDeliveryTag(), true);
            }
        }
        echo "done\n";
        return $channel->basic_ack($message->getDeliveryTag());
    }
);

while ($channel->is_consuming()) {
    $channel->wait();
}

File: mq/rabbitmq/py/broadcast/subscriber_printer.py
#!/usr/bin/env python3
#
# Copyright (C) 2019 jayson
#
import json
import random
import logging
import utils.logging
from utils.connection import get_exchange_connection
from config import RABBITMQ_MASTER
from pika.adapters.blocking_connection import BlockingChannel
from pika.spec import Basic, BasicProperties


def consumer(ch: BlockingChannel, method: Basic.Deliver, properties: BasicProperties, body: bytes):
    data = json.loads(body.decode('utf-8'))
    utils.logging.info("receive message <{}@{}>: '{}' from {}".format(
        data['producer'], data['index'], data['body'], data['from']))

    for _ in range(random.randint(1, (data['index'] % 5) + 1)):
        print('.', end='', flush=True)
        if random.randint(1, 25) == 1:
            utils.logging.warning('reject message <{}@{}>'.format(data['producer'], data['index']))
            print('rejected', flush=True)
            return ch.basic_reject(delivery_tag=method.delivery_tag)
    print('done', flush=True)
    ch.basic_ack(delivery_tag=method.delivery_tag)


if __name__ == '__main__':
    utils.logging.init_logging(logging.DEBUG)
    with get_exchange_connection(RABBITMQ_MASTER, 'logs_broadcast', 'fanout') as channel:
        channel.basic_qos(prefetch_count=1)
        channel.queue_declare(queue='logs_broadcast_printer', durable=True, auto_delete=True)
        channel.queue_bind(queue='logs_broadcast_printer', exchange='logs_broadcast')

        channel.basic_consume(queue='logs_broadcast_printer', on_message_callback=consumer)
        channel.start_consuming()
